sdk/forms/src/protected-view/iframe/index.ts

```diff
@@ -0,0 +1,125 @@
+import { renderView } from './components/iframe';
+import { type Logger, newLogger } from '../../common/logger';
+import { newSenderToSource, type Sender } from '../../common/events';
+import { ViewIframeEventValidator } from '../../common/models';
+import { ObjectFields, VaultClient } from '@piiano/vault-client';
+import { ViewIframeOptions } from '../../options';
+
+let allowUpdates = false;
+let fetchObjectsOptions: Omit<ViewIframeOptions, 'debug' | 'dynamic' | 'style'> | undefined = undefined;
+let initialized = false;
+let objects: Promise<ObjectFields[]> | undefined = undefined;
+let ready: Promise<void> | undefined = undefined;
+
+let sendToParent: Sender = () => {};
+let log: Logger = () => {};
+
+window.onmessage = ({ source, data }) => {
+  // TODO: when the iframe will be loaded from the vault then we should allow users to configure the allowed origins.
+  //  This will validate the form allow messages only from the client site origin and not from other iframes in the page or a parent iframe.
+  // const parentOrigin = 'http://localhost:3000';
+  // // allow messages from the parent only
+  // if (origin !== parentOrigin) return;
+
+  // We already validate the init options in the parent and return an indicative message there.
+  // The validation here is done for security reasons to prevent malicious messages with prototype pollution or other attacks.
+  // This is why the error message here doesn't have to be very descriptive as it's not a flow that should ever happen.
+  if (!ViewIframeEventValidator.parse(data)) {
+    return sendToParent('error', {
+      type: 'invalid-event',
+      message: 'Invalid event data.',
+    });
+  }
+
+  const { event } = data;
+
+  log(`Received "${event}" event from parent`);
+  switch (event) {
+    case 'init':
+      log = newLogger('iframe', data.payload.debug);
```

Comment:
I think it makes sense to move `log` assignment below initialized check (after line 49), because if it was already initialised - log variable should be also non-empty. No need to reassign it. It's not critical, so up to you.

Reply:
Actually to create create the `sendToParent` function we pass it the logger:
`sendToParent = newSenderToSource(source, log);`
Then in the initialization check we use `sendToParent` to report error if the view is already initialized:
`if (initialized) return sendToParent('error', { type: 'initialization', message: 'View already initialized' });`

So `log` must be constructed in the beginning...